Add bookings

egornowik21 opened the pull request.

egornowik21 committed:
проверка поиска

egornowik21 committed:
проверка поиска v2

egornowik21 committed:
Не работает метод patch для букинга

egornowik21 committed:
Неверно ставится время при запросе через State: Past

egornowik21 committed:
версия для ревью

pattyku commented on src/main/java/ru/practicum/shareit/booking/dao/BookingRepository.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+public interface BookingRepository extends JpaRepository<Booking, Long> {
+    @Query(value = " select * from bookings where booker_id = ?1", nativeQuery = true)
```
Почему используешь нативные запросы? Мы как раз с помощью JPA избавляемся от их написания
`List<Booking> findByBooker_Id(Long bookerId, Sort sort);`

egornowik21 replied:
да сложность была посмотреть, как выглядит по ним код. но я разобрался - исправляю

pattyku commented on src/main/java/ru/practicum/shareit/booking/dao/BookingRepository.java:
```diff
@@ -0,0 +1,22 @@
+    @Query(value = " select * from bookings as b join items as i on " +
+            "b.item_id = i.id " +
+            "where i.owner_id = ?1", nativeQuery = true)
+    List<Booking> findByOwnerId(Long userId);
```
И здесь можно обойтись без нативных запросов
`List<Booking> findByItem_Owner_Id(Long ownerId, Sort sort);`

egornowik21 replied:
поправил

pattyku commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,177 @@
+        try {
+            List<Booking> returnList = getListBookingByState(State.valueOf(state), bookingByUser);
+        } catch (IllegalArgumentException e) {
+            throw new ValidationException("Unknown state: UNSUPPORTED_STATUS");
```
Используешь try-catch не по назначению. Лучше в контроллере сделать проверку и не вызывать метод сервиса если нет такого статуса 

```
	@GetMapping
	public List<Booking> getAllBookingsByUserID(........) {
		State state = State.from(state)
				.orElseThrow(() -> new IllegalArgumentException("Unknown state: " + stateParam));
		return ..........;
	}
```

Только для этого нужно в enum добавить метод получения значения из строки

pattyku replied:
Почему здесь ничего не поправил? Проверку валидации статуса перенеси или в контроллер или в начале метода. У нас не поддерживаемы статус, мы идем в базу вытаскиваем пользователя, потом по нему список заказов и только после у нас все проверяется. Это какая колоссальная нагрузка на БД, чтоб выдать ошибку. Для проверки ничего из выше перечисленного не нужно, можно это сделать сразу как получили запрос

pattyku commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,177 @@
+    private List<Booking> getListBookingByState(State state, List<Booking> bookingByUser) {
+        List<Booking> bookingByState = new ArrayList<>();
+        switch (state) {
+            case ALL:
```
Сделал универсальный запрос в БД и потом фильтруешь. Чем плох такой способ. Вытягиваешь из БД всю информацию. А представь у тебя там очень много записей и вместо 10 завершенных заказов, вытягиваем из базы 1000 заказов. Лучше для каждого статуса делать отдельный запрос в репозитории
Пару примеров

case REJECTED (как раз пример как можно писать запросы к базе не используя нативный sql)
```
	@Query(" select b " +
			"from Booking b " +
			"where b.item.owner.id = ?1 " +
			"  and (b.status = 'CANCELED' or b.status = 'REJECTED' )")
	List<Booking> findOwnerCanceled(Long bookerId, Sort sort);
```

case FUTURE:
	`List<Booking> findByItem_Owner_IdAndStartIsAfter(Long bookerId, LocalDateTime start, Sort sort);`

pattyku commented on src/main/java/ru/practicum/shareit/item/dao/ItemRepository.java:
```diff
@@ -0,0 +1,20 @@
+    List<Item> search(String text);
+}
+
+/*@Query(" select i from Item i " +
```
Закомментированный код перед отправкой на ПР лучше удалять

egornowik21 replied:
убрал

pattyku commented on src/main/java/ru/practicum/shareit/item/dto/ItemDtoWithDate.java:
```diff
@@ -0,0 +1,34 @@
+    @NotBlank(message = "Имя не может быть пустым")
+    @NotNull(message = "Имя не может быть пустым")
+    @NotEmpty(message = "Описание не может быть пустым")
+    String name;
```
Для строк используется аннотация @NotBlank, остальные нужно удалить

egornowik21 replied:
готово

pattyku commented on src/main/java/ru/practicum/shareit/item/dto/ItemDtoWithDate.java:
```diff
@@ -0,0 +1,34 @@
+    @NotNull(message = "Описание не может быть пустым")
+    @NotEmpty(message = "Описание не может быть пустым")
+    String description;
+    @NotBlank
```
@NotNull вместо @NotBlank

egornowik21 replied:
на Boolean ставим notNull?

pattyku replied:
Да, все верно. Используешь класс Boolean. Он может быть null, но ни как не может быть пустым
Вот хорошая статься по данным аннотациям https://www.baeldung.com/java-bean-validation-not-null-empty-blank

pattyku commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -3,60 +3,216 @@
+        List<ItemDtoWithDate> returnItemList = new ArrayList<>();
+        List<Item> usersItems = itemRepository.findByOwnerIdOrderByIdAsc(userId);
+        for (Item item : usersItems) {
+            List<CommentDto> commentByUser = commentRepository.findByItem_id(item.getId()).stream()
```
Не обращаемся к репозиториям внутри цикла. Так производим n лишних запроса к базе данных. Получения комментариев нужно реализовать вне цикла. Это же относится ко всем остальным подобным кейсам

Идея для реализации, можно сделать и по другому
```
// Выгружаем комментарии, группируем их по вещам и конвертируем в dto
Map<Item, List<Comment>> comments = commentRepository.findByItemIn(items, Sort.by(DESC, "created"))
				.stream()
				.collect(groupingBy(Comment::getItem, toList()));
```

pattyku replied:
Вот этот пункт точно нужно исправить

pattyku commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -14,11 +13,20 @@
-    @Email(message = "Не является почтовым адресом")
-    @NotBlank(message = "Почтовый адрес не может быть пустым")
+    @Column(name="email",nullable = false,length = 320)
+    @NotBlank
```
в @Entity не нужно использовать аннотации валидаций @NotBlank @Email, не делаем здесь никаких проверок

egornowik21 replied:
у меня поступает на вход в контроллер entity, поэтому эти поля валидируются

pattyku replied:
Вот это плохо. Контроллер ничего не должен знать о модели. Все общение между контроллером и сервисом через dto. Везде используешь dto, а тут решил entity

pattyku reviewed (commented):
Есть комментарии, которые относятся ко всему коду

egornowik21 committed:
Сделано:
1) заменил методы с нативными запросами на запросные
2) Убрал try-catch и сделал валидацию статуса в switch-case
3) Переделал методы getAllBookingsByUserID и getAllBookingsByOwnerID
4) убрал закомментированный код
5) поменял валидацию строк на NotBlank

pattyku commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDtoInput.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.time.LocalDateTime;
+
+/**
```
Такие комментарии TODO лучше удалить по всему коду

pattyku commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -27,4 +28,5 @@ public class ItemDto {
     String description;
     @NotBlank
```
@NotNull

pattyku commented on src/main/java/ru/practicum/shareit/item/mapper/ItemMapper.java:
```diff
@@ -2,7 +2,9 @@
 import ru.practicum.shareit.item.model.Item;
+import ru.practicum.shareit.user.model.User;
 
 @Component
```
Для маппера не нужна аннотация @Component. Это утилитарный класс, который содержит статические методы и не имеет своего состояния. Для мапперов можно использовать аннотацию @NoArgsConstructor(access = AccessLevel.PRIVATE), чтобы создать приватный конструктор без параметров, тем самым запретив создание объектов этого класса
Комментарий относится ко всем мапперам

pattyku commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -16,11 +15,27 @@
 @FieldDefaults(level = AccessLevel.PRIVATE)
 @Data
```
Аннотация `@Data` не корректно работает с jpa, поэтому в @Entity ее не используем
https://habr.com/ru/company/haulmont/blog/564682/

egornowik21 committed:
Сделано:
1) Проверка статус на уровне контроллера
2) Убрал ToDo и все закомментированные места кода
3) Исправил метод по получению всех вещей по ID владельца
4) Убрал валидацию у entity и поменял возвращаемое значение методов контроллера и сервиса на DTO
5) Убрал DATA у entity и добавил к ней Getter и Setter
6) Поправил мапперы: заменил все аннотации component на конструкторы без парметров
7) Поправил чек стайл: убрал лишние импорты и отступы

egornowik21 committed:
checkstyle v1

pattyku commented on src/main/java/ru/practicum/shareit/booking/model/State.java:
```diff
@@ -0,0 +1,25 @@
+    WAITING,
+    REJECTED;
+
+    public static State fromString(String parameterName) {
```
Здесь можно было бы обернуть в Optional, например так
```
	public static Optional<BookingState> from(String stringState) {
		for (BookingState state : values()) {
			if (state.name().equalsIgnoreCase(stringState)) {
				return Optional.of(state);
			}
		}
		return Optional.empty();
	}
```

И тогда в контролле можно было бы сделать проверку в одну строку

```
BookingState state = BookingState.from(stateParam)
	.orElseThrow(() -> new IllegalArgumentException("Unknown state: " + stateParam));
```

Можно не исправлять, чтоб знал как можно сделать по другому
